cmd/derper/derper.go

```diff
@@ -254,39 +274,27 @@ func serveSTUN() {
 		log.Fatalf("failed to open STUN listener: %v", err)
 	}
 	log.Printf("running STUN server on %v", pc.LocalAddr())
+	serverSTUNListener(context.Background(), pc.(*net.UDPConn))
+}
 
+func serverSTUNListener(ctx context.Context, pc *net.UDPConn) {
+	var buf [64 << 10]byte
 	var (
-		stats           = new(metrics.Set)
-		stunDisposition = &metrics.LabelMap{Label: "disposition"}
-		stunAddrFamily  = &metrics.LabelMap{Label: "family"}
-
-		stunReadError  = stunDisposition.Get("read_error")
-		stunNotSTUN    = stunDisposition.Get("not_stun")
-		stunWriteError = stunDisposition.Get("write_error")
-		stunSuccess    = stunDisposition.Get("success")
-
-		stunIPv4 = stunAddrFamily.Get("ipv4")
-		stunIPv6 = stunAddrFamily.Get("ipv6")
+		n   int
+		ua  *net.UDPAddr
+		err error
 	)
-	stats.Set("counter_requests", stunDisposition)
-	stats.Set("counter_addrfamily", stunAddrFamily)
-	expvar.Publish("stun", stats)
-
-	var buf [64 << 10]byte
 	for {
-		n, addr, err := pc.ReadFrom(buf[:])
+		n, ua, err = pc.ReadFromUDP(buf[:])
```

Comment:
apparently this still escapes, probably from the pc.WriteTo on line 314. bah.

More reason to get netaddr in std, to make UDP more efficient in Go.

Reply:
Fascinating: https://blog.filippo.io/efficient-go-apis-with-the-inliner/
Found this while reading the code to see why `UDPConn.ReadFromUDP` didn't automatically allocate a `UDPAddr`.

Reply:
Lame. So `UDPConn.WriteToUDP` only escapes the input `addr` because it's returned in the error message. I tried doing the same function outlining trick, but can't seem to get the outer function to be inlineable since calling a function to do the error handling exceeds the inline budget.

Reply:
How about constructing a new, identical *net.UDPAddr to return in the error case? Something like:

```go
	if !ok {
                addrClone := *a
		return 0, &OpError{Op: "write", Net: c.fd.net, Source: c.fd.laddr, Addr: &addrClone, Err: syscall.EINVAL}
	}
```

I think then it'll only allocate on the actual error path.

If that doesn't work, there may be other things we can try. Ping me and we can pair on this. :)

Reply:
Oh, I see that you suggested that same thing lower down. :) Anyway, I think that that is a good idea. I assume you're on it, but let me know if you'd like my help.